### DISCUSSION TITLE
Best way to use functions to avoid code duplication

### DISCUSSION THREAD

**Discussion by LemonPie22** [filamentphp/filament#11459] [Help]
### Package

Panel builder

### Package Version

v3.2.24

### How can we help you?

I have a function to check for related parts before allowing a quote to be deleted:
```
public static function checkForRelatedQuotes(Part $record): bool
{
    if ($record->quotes()->exists()) {
        Notification::make()
            ->danger()
            ->title('Cannot delete part')
            ->body('Part #'.$record->item_number.
                ' cannot be deleted as it is a line item in quote #'.
                $record->quotes()->first()->id)
            ->actions([
                Action::make('view')
                    ->label('View quote')
                    ->button()
                    ->url(url('/').'/admin/quotes/'.$record->quotes()->first()->id)
            ])
            ->persistent()
            ->send();
        return true;
    }
    return false;
}
```
I want to call this from several places:

- Tables\Actions\DeleteAction
- Tables\Actions\DeleteBulkAction
- In Actions\DeleteAction on the part edit page

I can think of several ways to achieve this but I'd like to follow best practice.

My first approach was to add the function to my part resource and then call it by:
```
Tables\Actions\DeleteAction::make()
    // Check for related quotes and, if found, halt deletion
    ->before(function (Tables\Actions\DeleteAction $action, Part $record) {
        if (PartResource::checkForRelatedQuotes($record)) {
            $action->cancel();
        }
    }),
```
Or:
```
Tables\Actions\DeleteAction::make()
    // Check for related quotes and, if found, halt deletion
    ->before(function (Tables\Actions\DeleteAction $action, Part $record) {
        if (self::checkForRelatedQuotes($record)) {
            $action->cancel();
        }
    }),
```
The second option (using `self::` instead of specifically naming the resource) seems simpler but doesn't work from the edit page and, I believe, guarantees a new instantiation of the resource object every time it's called.

I considered moving my function to the `Part` model or to a service class. There are probably dozens of different ways to achieve this. Could someone point me in the right direction please?

**Comment by leandrocfe** [comment]
Maybe a trait?
